native/src/main/java/io/ayash/blockingqueue/BlockingQueue.java

```diff
@@ -0,0 +1,165 @@
+package io.ayash.blockingqueue;
+
+import java.util.Objects;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class BlockingQueue {
+    public final Object[] queue;
+    private final int capacity;
+    private final ReentrantLock lock = new ReentrantLock(true);
+    private final Condition notFull = lock.newCondition();
+    private final Condition notEmpty = lock.newCondition();
+    private int head = 0;
+    private int tail = 0; // next available slot
+    private int size = 0;
+
+    public BlockingQueue(int capacity) {
+        this.capacity = capacity;
+        queue = new Object[capacity];
+    }
+
+    public void put(Object item) throws InterruptedException {
+        Objects.requireNonNull(item);
+        lock.lockInterruptibly();
+        try {
+            while (size == capacity) {
+                notFull.await();
+            }
+            queue[tail] = item;
+            tail = (tail + 1) % capacity;
+            size++;
+            notEmpty.signal();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+
+    public Object take() throws InterruptedException {
+        lock.lockInterruptibly();
+        try {
+            while (size == 0) {
+                notEmpty.await();
+            }
+            Object item = queue[head];
+            head = (head + 1) % capacity;
+            size--;
+            notFull.signal();
+            return item;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public boolean add(Object item) {
+        Objects.requireNonNull(item);
```

Comment:
Shall we handle exception here? When there is an exception you may need to convert and return it as a `BError`